### PR TITLE
feat(ci): Upload test results to Datadog

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

A couple of notes:

- This should be reporting skipped tests, and `nextest` appears to support this in the junit output but I don't see any of our skipped tests in Datadog
- This is skipping reporting on Windows test runs as I got bogged down in trying to cross-platform a shellscript and get node installed on Windows
- We should replace the API key used here as it's currently a test key I created a while back

Results can be found [here](https://app.datadoghq.com/ci/test-runs) in our datadog org

### CODE REVIEW

File: scripts/upload-test-results.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# SUMMARY
+#
+#   Upload `cargo-nextest` JUnit output to Datadog
+
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+set -x
+
+DD_ENV="${DD_ENV:-"local"}" datadog-ci junit upload \
+  --service vector \
+  target/nextest/default/junit.xml
```
Comment (jszwedko) [inline scripts/upload-test-results.sh:16]: Would it be possible to tag these to separate out the OS / platform results? I do see in the UI that there is a section but I'm not sure where it gets that data:

Screen Shot 2022-03-09 at 10 02 54 AM

Comment (spencergilbert) [reply scripts/upload-test-results.sh:16]: Yeah, should be able to: https://docs.datadoghq.com/continuous_integration/setup_tests/junit_upload/#collecting-environment-configuration-metadata

Comment (spencergilbert) [reply scripts/upload-test-results.sh]: Added tags for arch and platform (`uname -m` and `uname -s` based)

File: .config/nextest.toml [hunk 2]
```diff
@@ -1,2 +1,18 @@
 [profile.default]
+# "retries" defines the number of times a test should be retried. If set to a
```
Comment (jszwedko) [inline .config/nextest.toml:2]: 👍 thanks for the comments.

File: Makefile [hunk 3]
```diff
@@ -297,6 +297,9 @@ target/%/vector.tar.gz: target/%/vector CARGO_HANDLES_FRESHNESS
 .PHONY: test
 test: ## Run the unit test suite
 	${MAYBE_ENVIRONMENT_EXEC} cargo nextest run --workspace --no-fail-fast --no-default-features --features "${DEFAULT_FEATURES}" ${SCOPE}
+ifneq ($(OPERATING_SYSTEM), Windows)
+	@scripts/upload-test-results.sh
+endif
```
Comment (jszwedko) [inline Makefile:302]: Might be worth leaving a comment about why we don't run this on Windows for future readers.

**Review by jszwedko — APPROVED** [review]
Nice! I'm excited to have this integrated.

File: Makefile (hunk 3 above)
Comment (spencergilbert) [reply Makefile:302]: Fair - more usable than just the github issue,

File: .config/nextest.toml (hunk 2 above)
Comment (spencergilbert) [reply .config/nextest.toml:2]: NP, I thought especially that one was useful because of how flakey tests are reported

**Pull request merged**